subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -78,12 +81,19 @@ public JavaBasePlugin(Instantiator instantiator, ObjectFactory objectFactory) {
         this.objectFactory = objectFactory;
     }
 
-    public void apply(ProjectInternal project) {
+    public void apply(final ProjectInternal project) {
         project.getPluginManager().apply(BasePlugin.class);
         project.getPluginManager().apply(ReportingBasePlugin.class);
 
-        JavaPluginConvention javaConvention = new JavaPluginConvention(project, instantiator);
+        JavaPluginConvention javaConvention = DeprecationLogger.whileDisabled(new Factory<JavaPluginConvention>() {
+            @Override
+            public JavaPluginConvention create() {
+                return new JavaPluginConvention(project, instantiator);
+            }
+        });
         project.getConvention().getPlugins().put("java", javaConvention);
+        project.getExtensions().add("sourceSets", javaConvention.getSourceSets());
```

Comment:
:x: this should declare `SourceSetContainer` as a public type otherwise it'll be inferred and will be `DefaultSourceSetContainer`, exposing internals